src/alpaca_codegen.erl

```diff
@@ -290,17 +291,52 @@ gen_expr(Env, #alpaca_apply{expr={symbol, _Line, Name}, args=[{unit, _}]}) ->
                 1 -> cerl:c_fname(list_to_atom(Name), 1)
             end,
     {Env, cerl:c_apply(FName, [cerl:c_atom(unit)])};
-gen_expr(Env, #alpaca_apply{expr={symbol, _Line, Name}, args=Args}) ->
-    FName = case proplists:get_value(Name, Env#env.module_funs) of
-                undefined ->
-                    cerl:c_var(list_to_atom(Name));
-                Arity ->
-                    cerl:c_fname(list_to_atom(Name), Arity)
-            end,
-    Apply = cerl:c_apply(
-              FName, 
-              [A || {_, A} <- [gen_expr(Env, E) || E <- Args]]),
-    {Env, Apply};
+gen_expr(Env, #alpaca_apply{expr={symbol, L, Name}=FExpr, args=Args}) ->
+    DesiredArity = length(Args),
+    {FName, Curry, Arity} = case proplists:get_all_values(Name, Env#env.module_funs) of
+        [] -> {cerl:c_var(list_to_atom(Name)), false, 0};
+        AvailFuns ->
+            %% If we have an exact arity match, use that, otherwise curry
+            case lists:filter(fun(X) -> X =:= DesiredArity end, AvailFuns) of
+                [A] -> {cerl:c_fname(list_to_atom(Name), A), false, A};
+                _ ->  
+                    PossibleCurries = lists:filter(fun(X) -> X > DesiredArity end, AvailFuns),
+                    %% The typer ensures that we can curry unambiguously
+                    case PossibleCurries of
```

Comment:
Not strictly necessary with a case here, as as you say the typer already guarantees it will be an unambiguous match. Hence: 
````
%% The typer ensures that we can curry unambiguously
[CurryArity] = lists:filter(fun(X) -> X > DesiredArity end, AvailFuns),
{cerl:c_fname(list_to_atom(Name), CurryArity), true, CurryArity}
````